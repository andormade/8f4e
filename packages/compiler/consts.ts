export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
