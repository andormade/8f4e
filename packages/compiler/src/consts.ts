export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
export const LOGIC_HIGH = I16_SIGNED_LARGEST_NUMBER;
export const LOGIC_LOW = 0;
