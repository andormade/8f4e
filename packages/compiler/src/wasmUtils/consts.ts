export const WASM_MEMORY_PAGE_SIZE = 65536;
