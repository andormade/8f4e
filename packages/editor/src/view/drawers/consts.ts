export const VGRID = 8;
export const HGRID = 16;
