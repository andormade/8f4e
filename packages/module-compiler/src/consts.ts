export const WORD_LENGTH = 4;
