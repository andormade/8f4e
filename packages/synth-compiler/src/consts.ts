export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I32_SIGNED_LARGEST_NUMBER = 2147483647;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
export const LOGIC_HIGH = I16_SIGNED_LARGEST_NUMBER;
export const LOGIC_LOW = 0;
export const WORD_LENGTH = 4;
