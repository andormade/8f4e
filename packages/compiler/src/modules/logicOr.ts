import {
	Instruction,
	Type,
	br,
	createFunctionBody,
	i32const,
	i32load,
	i32store,
	ifelse,
} from '../../../byteCodeUtils/src';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER } from '../consts';

export enum Memory {
	ZERO,
	INPUT_1_POINTER,
	INPUT_2_POINTER,
	OUTPUT,
}

const or: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(offset(Memory.INPUT_1_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				...i32const(offset(Memory.OUTPUT)),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				...i32store(),
				...br(1),
			]),

			...i32const(offset(Memory.INPUT_2_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				...i32const(offset(Memory.OUTPUT)),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				...i32store(),
				...br(1),
			]),

			...i32const(offset(Memory.OUTPUT)),
			...i32const(0),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset: offset(0),
		initialMemory: [0, offset(Memory.ZERO), offset(Memory.ZERO), 0],
		memoryAddresses: [
			{ address: offset(Memory.OUTPUT), id: 'out' },
			{
				address: offset(Memory.INPUT_1_POINTER),
				id: 'in:1',
				default: offset(Memory.ZERO),
			},
			{
				address: offset(Memory.INPUT_2_POINTER),
				id: 'in:2',
				default: offset(Memory.ZERO),
			},
		],
	};
};

export default or;
