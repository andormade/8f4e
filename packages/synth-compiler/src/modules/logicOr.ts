import { Instruction, Type, br, createFunctionBody, i32const, i32load, i32store, ifelse } from '@8f4e/bytecode-utils';
import { MemoryTypes, ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER } from '../consts';

export enum Memory {
	ZERO,
	INPUT_1_POINTER,
	INPUT_2_POINTER,
	OUTPUT,
}

const or: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(offset.byte(Memory.INPUT_1_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				...i32const(offset.byte(Memory.OUTPUT)),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				...i32store(),
				...br(1),
			]),

			...i32const(offset.byte(Memory.INPUT_2_POINTER)),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.VOID, [
				...i32const(offset.byte(Memory.OUTPUT)),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				...i32store(),
				...br(1),
			]),

			...i32const(offset.byte(Memory.OUTPUT)),
			...i32const(0),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		byteAddress: offset.byte(0),
		wordAddress: offset.word(0),
		memoryMap: [
			{ type: MemoryTypes.PRIVATE, address: Memory.ZERO, default: 0 },

			{
				type: MemoryTypes.INPUT_POINTER,
				address: offset.byte(Memory.INPUT_1_POINTER),
				id: 'in:1',
				default: offset.byte(Memory.ZERO),
			},
			{
				type: MemoryTypes.INPUT_POINTER,
				address: offset.byte(Memory.INPUT_2_POINTER),
				id: 'in:2',
				default: offset.byte(Memory.ZERO),
			},

			{ type: MemoryTypes.OUTPUT, address: offset.byte(Memory.OUTPUT), id: 'out', default: 0 },
		],
	};
};

export default or;
