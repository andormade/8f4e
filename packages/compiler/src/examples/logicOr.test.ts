import logicOr from './logicOr.asm';
import { I16_SIGNED_LARGEST_NUMBER } from './consts';

import { createTestModule, TestModule } from '../';

let testModule: TestModule;

const fixtures: [input1: number, input2: number, output: number][] = [
	[10, 10, I16_SIGNED_LARGEST_NUMBER],
	[0, 10, I16_SIGNED_LARGEST_NUMBER],
	[10, 0, I16_SIGNED_LARGEST_NUMBER],
	[0, 0, 0],
	[-10, 0, 0],
];

describe('logicOr', () => {
	beforeAll(async () => {
		testModule = await createTestModule(logicOr);
	});

	beforeEach(() => {
		testModule.reset();
	});

	test('if the generated ast matches with the snapshot', () => {
		expect(testModule.ast).toMatchSnapshot();
	});

	test('if the wat code matches with the snapshot', () => {
		expect(testModule.wat).toMatchSnapshot();
	});

	test('if the generated memory map matches with the snapshot', () => {
		expect(testModule.memoryMap).toMatchSnapshot();
	});

	test.each(fixtures)('given inputs %p and %p, the expected output is %p', (input1, input2, output) => {
		const { memory, test } = testModule;

		const in1 = memory.allocMemoryForPointer('in:1');
		const in2 = memory.allocMemoryForPointer('in:2');
		memory.set(in1, input1);
		memory.set(in2, input2);
		test();
		expect(memory.get('out')).toBe(output);
	});
});
