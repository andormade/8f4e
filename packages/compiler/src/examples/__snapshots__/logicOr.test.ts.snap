// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logicOr if the generated ast matches with the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "logucOr",
      },
    ],
    "instruction": "module",
    "lineNumber": 0,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "defaultValue",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "int",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in:1",
      },
      Object {
        "type": "identifier",
        "value": "&defaultValue",
      },
    ],
    "instruction": "int*",
    "lineNumber": 3,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "in:2",
      },
      Object {
        "type": "identifier",
        "value": "&defaultValue",
      },
    ],
    "instruction": "int*",
    "lineNumber": 4,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "int",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "HIGH",
      },
      Object {
        "type": "identifier",
        "value": "I16_SIGNED_LARGEST_NUMBER",
      },
    ],
    "instruction": "const",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "*in:1",
      },
    ],
    "instruction": "push",
    "lineNumber": 9,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 10,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterThan",
    "lineNumber": 11,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 12,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 13,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "HIGH",
      },
    ],
    "instruction": "push",
    "lineNumber": 14,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 15,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "branch",
    "lineNumber": 16,
  },
  Object {
    "arguments": Array [],
    "instruction": "ifEnd",
    "lineNumber": 17,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "*in:2",
      },
    ],
    "instruction": "push",
    "lineNumber": 19,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 20,
  },
  Object {
    "arguments": Array [],
    "instruction": "greaterThan",
    "lineNumber": 21,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "void",
      },
    ],
    "instruction": "if",
    "lineNumber": 22,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&out",
      },
    ],
    "instruction": "push",
    "lineNumber": 23,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "HIGH",
      },
    ],
    "instruction": "push",
    "lineNumber": 24,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 25,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 1,
      },
    ],
    "instruction": "branch",
    "lineNumber": 26,
  },
  Object {
    "arguments": Array [],
    "instruction": "ifEnd",
    "lineNumber": 27,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "out",
      },
    ],
    "instruction": "push",
    "lineNumber": 29,
  },
  Object {
    "arguments": Array [
      Object {
        "isInteger": true,
        "type": "literal",
        "value": 0,
      },
    ],
    "instruction": "push",
    "lineNumber": 30,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 31,
  },
  Object {
    "arguments": Array [],
    "instruction": "moduleEnd",
    "lineNumber": 32,
  },
]
`;

exports[`logicOr if the generated memory map matches with the snapshot 1`] = `
Map {
  "defaultValue" => Object {
    "alignmentAdjustedSize": 1,
    "byteAddress": 0,
    "default": 0,
    "id": "defaultValue",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "type": "int",
    "wordAddress": 0,
    "wordSize": 4,
  },
  "in:1" => Object {
    "alignmentAdjustedSize": 1,
    "byteAddress": 4,
    "default": 0,
    "id": "in:1",
    "isInteger": true,
    "isPointer": true,
    "isPointingToInteger": true,
    "isPointingToPointer": false,
    "type": "int*",
    "wordAddress": 1,
    "wordSize": 4,
  },
  "in:2" => Object {
    "alignmentAdjustedSize": 1,
    "byteAddress": 8,
    "default": 0,
    "id": "in:2",
    "isInteger": true,
    "isPointer": true,
    "isPointingToInteger": true,
    "isPointingToPointer": false,
    "type": "int*",
    "wordAddress": 2,
    "wordSize": 4,
  },
  "out" => Object {
    "alignmentAdjustedSize": 1,
    "byteAddress": 12,
    "default": 0,
    "id": "out",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "type": "int",
    "wordAddress": 3,
    "wordSize": 4,
  },
}
`;

exports[`logicOr if the wat code matches with the snapshot 1`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    (local i32 i32)
    i32.const 4
    i32.load
    i32.load
    i32.const 0
    i32.gt_s
    if  ;; label = @1
      i32.const 12
      i32.const 32767
      i32.store
      br 1 (;@0;)
    end
    i32.const 8
    i32.load
    i32.load
    i32.const 0
    i32.gt_s
    if  ;; label = @1
      i32.const 12
      i32.const 32767
      i32.store
      br 1 (;@0;)
    end
    i32.const 12
    i32.load
    i32.const 0
    local.set 1
    local.set 0
    local.get 0
    i32.const 65535
    i32.gt_s
    if (result i32)  ;; label = @1
      i32.const 0
    else
      local.get 0
    end
    local.get 1
    i32.store)
  (export \\"test\\" (func 0)))
"
`;
